### PR TITLE
BUG: sparse.csgraph, array types: support non-zero `fill_value`s

### PR DESCRIPTION
**Pull request by mtsokol**
Connected to: https://github.com/scipy/scipy/issues/20763, https://github.com/pydata/sparse/pull/685, https://github.com/willow-ahrens/finch-tensor/pull/59

Hi @hameerabbasi,

This PR allows to pass `pydata/sparse` arrays with `np.inf` fill-values to the selected `scipy.sparse.csgraph` functions.

### CODE REVIEW

**Comment by mtsokol** [comment]
I think once `pydata/sparse` maintenance release is made then CI should be green.

**Comment by hameerabbasi** [comment]
The tests at `scipy/sparse/csgraph/tests/test_pydata_sparse.py` might need adjusting -- maintainence release is up on PyPI: https://pypi.org/project/sparse/0.15.4/

File: scipy/sparse/csgraph/_validation.py [hunk 1]
```diff
@@ -18,7 +18,12 @@ def validate_graph(csgraph, directed, dtype=DTYPE,
     if not (csr_output or dense_output):
         raise ValueError("Internal: dense or csr output must be true")
 
-    csgraph = convert_pydata_sparse_to_scipy(csgraph)
+    accept_fv = []
```
Comment (hameerabbasi) [inline scipy/sparse/csgraph/_validation.py:21]:
```suggestion
    accept_fv = [null_value_in]
```

File: scipy/sparse/csgraph/_validation.py [hunk 2]
```diff
@@ -18,7 +18,12 @@ def validate_graph(csgraph, directed, dtype=DTYPE,
...
+    if infinity_null:
+        accept_fv.append(np.inf)
+    if nan_null:
+        accept_fv.append(np.nan)
+    csgraph = convert_pydata_sparse_to_scipy(csgraph, accept_fv=accept_fv or None)
```
Comment (hameerabbasi) [inline scipy/sparse/csgraph/_validation.py:26]:
```suggestion
    csgraph = convert_pydata_sparse_to_scipy(csgraph, accept_fv=accept_fv)
```

File: scipy/sparse/csgraph/_validation.py (hunk 1 above)
Comment (mtsokol) [reply scipy/sparse/csgraph/_validation.py]: Done!

File: scipy/sparse/csgraph/_validation.py (hunk 2 above)
Comment (mtsokol) [reply scipy/sparse/csgraph/_validation.py]: Done!

**Review by hameerabbasi — APPROVED** [review]
Thanks for the work on this @mtsokol!

File: scipy/sparse/_sputils.py [hunk 3]
```diff
@@ -390,14 +390,16 @@ def is_pydata_spmatrix(m) -> bool:
...
+    target_format: Optional[Literal["csc", "csr"]] = None,
+    accept_fv: Any = None,
 ) -> Union[Any, "sp.spmatrix"]:
     """
     Convert a pydata/sparse array to scipy sparse matrix,
     pass through anything else.
     """
     if is_pydata_spmatrix(arg):
-        arg = arg.to_scipy_sparse()
+        arg = arg.to_scipy_sparse(accept_fv=accept_fv)
```
Comment (rgommers) [inline scipy/sparse/_sputils.py:402]: This looks like it's going to break for `sparse<0.15.4`.

**Review by rgommers — CHANGES_REQUESTED** [review]
The test and changes LGTM, except for relying on a feature that is brand new in pydata sparse. How about wrapping the `to_scipy_sparse(accept_fv)` call in a try-except, catching the `TypeError` and then callign again without use of `accept_fv`? That keeps the default behavior with `0` fill value supported for older versions.

File: scipy/sparse/_sputils.py (hunk 3 above)
Comment (mtsokol) [reply scipy/sparse/_sputils.py]: Done!

File: scipy/sparse/csgraph/_tools.pyx [hunk 4]
```diff
@@ -502,7 +503,9 @@ def reconstruct_path(csgraph, predecessors, directed=True):
 
     sctree = csr_matrix((data, indices, indptr), shape=(N, N))
     if is_pydata_sparse:
-        sctree = pydata_sparse_cls.from_scipy_sparse(sctree)
+        sctree = pydata_sparse_cls.from_scipy_sparse(
+            sctree, fill_value=pydata_sparse_fill_value
+        )
```
Comment (hameerabbasi) [inline scipy/sparse/csgraph/_tools.pyx:508]: The point about new `sparse` version also applies here

File: scipy/sparse/csgraph/_validation.py [hunk 5]
```diff
@@ -18,7 +18,12 @@ def validate_graph(csgraph, directed, dtype=DTYPE,
     if not (csr_output or dense_output):
         raise ValueError("Internal: dense or csr output must be true")
 
-    csgraph = convert_pydata_sparse_to_scipy(csgraph)
+    accept_fv = [null_value_in]
+    if infinity_null:
+        accept_fv.append(np.inf)
+    if nan_null:
+        accept_fv.append(np.nan)
+    csgraph = convert_pydata_sparse_to_scipy(csgraph, accept_fv=accept_fv)
```
Comment (hameerabbasi) [inline scipy/sparse/csgraph/_validation.py:26]: Also here.

**Review by hameerabbasi — COMMENTED** [review]
A few more places where a new `sparse` version is required.

File: scipy/sparse/csgraph/_validation.py (hunk 5 above)
Comment (mtsokol) [reply scipy/sparse/csgraph/_validation.py:26]: But that's our function, it isn't imported from `sparse`. Why would we need it in a `try: catch:`?

Comment (hameerabbasi) [reply scipy/sparse/csgraph/_validation.py:26]: Ah, that's my bad -- thanks for the context.

File: scipy/sparse/csgraph/_tools.pyx (hunk 4 above)
Comment (mtsokol) [reply scipy/sparse/csgraph/_tools.pyx]: Done!

File: scipy/sparse/csgraph/tests/test_pydata_sparse.py [hunk 6]
```diff
@@ -147,3 +147,37 @@ def test_min_weight_full_bipartite_matching(graphs):
...
+        spgraph.dijkstra,
+        spgraph.floyd_warshall,
+        spgraph.bellman_ford,
+        spgraph.johnson,
+        spgraph.minimum_spanning_tree,
+    ]
+)
+@pytest.mark.parametrize(
+    "fill_value, comp_func",
+    [(np.inf, np.isinf), (np.nan, np.isnan)],
```
Comment (hameerabbasi) [inline scipy/sparse/csgraph/tests/test_pydata_sparse.py:165]: Minor nit -- we should use `np.isposinf` or something. `np.isinf` allows both positive and negative infinities

Comment (hameerabbasi) [reply scipy/sparse/csgraph/tests/test_pydata_sparse.py:165]:
```suggestion
    [(np.inf, np.isposinf), (np.nan, np.isnan)],
```

**Review by hameerabbasi — APPROVED** [review]
Thanks for all the hard work here @mtsokol!

**Comment by rgommers** [comment]
Some comments above each try-except would be useful. I suggest:
```python
# The `accept_fv` keyword is new in PyData Sparse 0.15.4 (May 2024),
# remove the `except` once the minimum supported version is >=0.15.4
```

**Comment by mtsokol** [comment]
Sure! I added comments.

**Review by rgommers — APPROVED** [review]
LGTM now, let's give this a go. Thanks @mtsokol, and thanks @hameerabbasi for the reviews.

**Comment by WarrenWeckesser** [comment]
It looks like the new code doesn't handle `sparse` versions before 0.15.4 correctly: https://github.com/scipy/scipy/issues/20861

**Pull request merged**